### DISCUSSION TITLE
Command Handler

### DISCUSSION THREAD

**Discussion by ghost** [mojidev-py/discmoji#1] [Show and tell]
I've tested and found a way for you to handle your commands, as well as a decorator for it.

bot.py :
```py
class Bot:
  def __init__(self) -> None:
    self._all_cmds : List[Command] = [] # if you'd like to keep it iguess
    self.command : Command = Command
    self.command.bot : Self = self
```

command.py :
```py
class Command:
  def __init__(self, /, name : str = None) -> None: # arguments for the decorator
    self.name : str = name
    self.__error_handlers : List[Callable] = []

  def __call__(self, function : Callable | None = None) -> None:
    if not function:
      try:
        return self.callback()
      except:
        tasks = []
        for handler in self.__error_handlers:
          tasks.append(handler())
        asyncio.run(gather(*tasks))
    self.callback : Callable = function
    if not self.name: self.name : str = func.__name__
    self.bot._all_cmds += self
    return self # allows the function ( command object ) to still be used later on in the code

  def error(self, function) -> None:
    self.__error_handlers += function
```
` Command.__init__() ` accepts the arguments for the decorator :
```py
@bot.command()
async def test(ctx : Context) -> None: ...

@bot.command(name = "test 2")
async def test2(ctx : Context) -> None: ...

# name defaults to the function's name if not provided in the decorator
```
` Command.__call__() ` handles the callback for the command. When a ` function ` is passed ( or when the decorator is called / used ), it reforms the function into a ` Command ` object ( lastly done by ` return self ` ). Otherwise if ` function ` parameter is not passed ( or when the function ( command object ) is called just like any other regular function, it instead calls the function callback.
` Command.error() ` is a decorator that appends the ` function ` to the error handlers for the command and does not return anything, thus the dev cannot make use the function again when it is called or referenced further in the code.
```py
@bot.command()
async def test(context : Context) -> None: ...

@test.error
async def error(context : Context, error : Exception) -> None: ...

error() # <-- NoneType object
```

PS : I changed some lines of the code while writing this and is untested, so some stuff might break, or not, idk.

**Comment by mojidev-py** [comment]
tysm for the help! I credited you in the example file, and the command file, I'll ask if this works!
